### PR TITLE
Refactors uplinks to be components, fix various bugs

### PR DESCRIPTION
**Pull request by Cyberboss**
Replaced /obj/item/var/obj/item/device/uplink/hidden_uplink with /datum/component/uplink

Same functions just as a component instead of an abstract object, access it with GetComponent(/datum/component/uplink)

Fixes uplinks not printing out at the end of game if they are destroyed

Fixes some double "The"s in uplink code

Fixed cyber_implants uplink item organs not coming in a cyber_implants box (is there an issue for this?)

Fixes some other things that I forgot about

Set ui_host() for components to be the parent

Added COMSIG_PARENT_ATTACKBY, sent when attackby is called

Refactored purchase logging into a separate datum which is stored in GLOB.uplink_purchase_logs which is a lazy list, keyed by an uplink owner's key, of lists of /datum/uplink_purchase_log

### CODE REVIEW

File: code/game/objects/items/stacks/telecrystal.dm [hunk 1]
```diff
@@ -10,28 +10,27 @@
...
+	if(istype(I, /obj/item/weapon/cartridge/virus/frame))
 		var/obj/item/weapon/cartridge/virus/frame/cart = I
 		if(!cart.charges)
-			to_chat(user, "<span class='notice'>The [cart] is out of charges, it's refusing to accept the [src]</span>")
+			to_chat(user, "<span class='notice'>[cart] is out of charges, it's refusing to accept [src]</span>")
 			return
 		cart.telecrystals += amount
 		use(amount)
-		to_chat(user, "<span class='notice'>You slot [src] into the [cart].  The next time it's used, it will also give telecrystals</span>")
+		to_chat(user, "<span class='notice'>You slot [src] into [cart]. The next time it's used, it will also give telecrystals</span>")
```
Comment (Tacolizard) [inline code/game/objects/items/stacks/telecrystal.dm]: need a period at the end of this sentence

**Comment by Cyberboss** [comment]
>Needs testing

**Comment by optimumtact** [comment]
It's not our responsibility for you to organise testing and if you can't do it in a timely manner I'm not going to leave the PR sitting around for weeks

**Comment by Cyberboss** [comment]
```
<%Cyberboss> Someone throw on #29785
<TheGhostOfWhibyl1> https://github.com/tgstation/tgstation/pull/29785 - 'Refactors uplinks to be components, fix various bugs' by Cyberboss
<%Cyberboss> and #29268 when they get a chance
<TheGhostOfWhibyl1> https://github.com/tgstation/tgstation/pull/29268 - 'Port material containers to datum components' by Cyberboss
<Mr_Bagil> REPO: Test merging PR #29741...
<TheGhostOfWhibyl1> https://github.com/tgstation/tgstation/pull/29741 - 'Clockwork Cult Rework: Proof-of-concept' by Xhuis
```

**Comment by Cyberboss** [comment]
Time to code testmerge requests into the server tools reee

**Comment by optimumtact** [comment]
:+1:

**Comment by optimumtact** [comment]
@Cyberboss can you fix merge conflicts

**Comment by optimumtact** [comment]
Has this had a test merge?

**Comment by Cyberboss** [comment]
Yes

File: code/_onclick/item_attack.dm [hunk 2]
```diff
@@ -16,10 +16,10 @@
 
 // No comment
 /atom/proc/attackby(obj/item/W, mob/user, params)
-	return
+	return SendSignal(COMSIG_PARENT_ATTACKBY, W, user, params)
 
 /obj/attackby(obj/item/I, mob/living/user, params)
-	return I.attack_obj(src, user)
+	return ..() || I.attack_obj(src, user)
```
Comment (optimumtact) [inline code/_onclick/item_attack.dm]: does this work

File: code/datums/components/component.dm [hunk 3]
```diff
@@ -75,6 +75,9 @@
 	LAZYCLEARLIST(signal_procs)
 	return ..()
 
+/datum/component/ui_host()	//makes sense
```
Comment (optimumtact) [inline code/datums/components/component.dm]: this comment doesn't

File: code/_onclick/item_attack.dm (hunk 2 above)
Comment (Cyberboss) [reply code/_onclick/item_attack.dm]: A component can return TRUE from it's signal processor to stop attack animations. Yes

**Comment by lzimann** [comment]
#30170

**Comment by KorPhaeron** [comment]
@Cyberboss 

Conflicts

**Comment by optimumtact** [comment]
Merge this already

**Pull request closed without merging**